docs/_docs/05-ug-optimize.md

```diff
@@ -0,0 +1,91 @@
+Two or more index files can be merged with each other, if and only if they all have index records which belong to the
+same bucket (according to index configuration).
+When running `optimizeIndex` command, Hyperspace tries to find such groups of index files and merge them together.
+If all index records for each given bucket are already stored in a single index file, then there wont be any
+index files merge during optimize and physical layout of index files after running optimize will be the same as the
+original layout. An example of such an index is an index right after creation or full refresh.    
+
+Currently, there are two optimize modes available for an index: `"quick"` and `"full"`. These modes differ with each other
+in terms of the subset of index files they identify and try to merge.
+ 
+### Optimize Index - Quick Mode
+Using `optimizeIndex` command with the `"quick"` mode on an index with many index files, due to incremental index refresh,
+causes Hyperspace look for index files which are smaller than a configurable size threshold and try to merge them.
+This mode tries to achieve a moderate query performance improvement through a fast optimize index process.
+The size threshold for an index file to be eligible for merging during quick optimization can be changed.
+Check the [configuration](https://microsoft.github.io/hyperspace/docs/ug-configuration/) page to see how this threshold can be adjusted.
+Quick mode is the default mode for `optimizeIndex`.
+
+Assume you have an index with the name `empIndex`. After adding some data files to the dataset `empIndex` is created on
+and refreshing it in the incremental mode, you can optimize `empIndex` in the quick mode as below:
+
+Scala:
+```scala
+import com.microsoft.hyperspace._
+
+val hs = new Hyperspace(spark)
+hs.optimizeIndex("empIndex", "quick")
+``` 
+
+Python:
+
+```python
+from hyperspace import Hyperspace
+
+hs = Hyperspace(spark)
+hs.optimizeIndex("empIndex", "quick")
+```
+
+### Optimize Index - Full Mode
+In the `"full"` mode, `optimizeIndex` command considers all existing index files as candidates for merging. Therefore,
+Hyperspace does a full scan on current index content and identifies groups of index files which have index records belonging
+to the same bucket according to the index configuration. It then replaces each group with a single index file created through
+merging all the files in that group together. This mode tries to achieve the best query performance improvement via a
+potentially slow optimize index process.
+You should note that `optimizeIndex` differs from `refreshIndex` in the sense that `optimizeIndex` is an index-only operation.
+Unlike `refreshIndex`, when `optimizeIndex` is running on an index, it does not look at the current state of the source data files.
+If there were any source data file changes after index creation or last refresh, `optimizeIndex` will not apply those data changes
+to the index files.
```

Comment:
thanks @pirz , could you please move this to the beginning / introduction section of the page? Currently it shows up in "full mode" section. It might lead to minor confusion that it's something specific to full mode.

Reply:
Good point. Moved it. Thnx!